Add readiness health check for Kafka Streams (#5125)
quarkusio/quarkus: docs/src/main/asciidoc/kafka-streams.adoc

Now start Docker Compose as described above
(don't forget to rebuild the container images).

> == Kafka Streams Health Check

Reviewer: ```suggestion
== Kafka Streams Health Checks
```

If you are using the `quarkus-smallrye-health` extension, `quarkus-kafka-streams` will automatically add:

* a readiness health check to validate that all topics declared in the `quarkus.kafka-streams.topics` property are created,
* a liveness health check based on the Kafka Streams state.

> So when you access the `/health` endpoint of your application you will have information about the state of the KStreams and the available and/or missing topics.

Reviewer: ```suggestion
So when you access the `/health` endpoint of your application you will have information about the state of the Kafka Streams and the available and/or missing topics.
```

> This is an example of `DOWN` status:

Reviewer: ```suggestion
This is an example of when the status is `DOWN`:
```

[source, subs=attributes+]
----
curl -i http://aggregator:8080/health

HTTP/1.1 503 Service Unavailable
content-type: application/json; charset=UTF-8
content-length: 454

{
    "status": "DOWN",
    "checks": [
        {
            "name": "Kafka Streams state health check",  <1>
            "status": "DOWN",
            "data": {
                "state": "CREATED"
            }
        },
        {

>             "name": "Kafka Streams topic(s) health check",  <2>

Reviewer: ```suggestion
            "name": "Kafka Streams topics health check",  <2>
```
I will get rid of the already existing ones with `(s)`. It's less readable and doesn't bring much.

            "status": "DOWN",
            "data": {
                "available_topics": "weather-stations,temperature-values",
                "missing_topics": "hygrometry-values"
            }
        }
    ]
}
----
<1> Liveness health check. Also available at `/health/live` endpoint.
<2> Readiness health check. Also available at `/health/ready` endpoint.

> So as you can see, the status is `DOWN` as soon as one of `quarkus.kafka-streams.topics` is missing or the Kafka Streams `state` is not `RUNNING`.

Reviewer: ```suggestion
So as you can see, the status is `DOWN` as soon as one of the `quarkus.kafka-streams.topics` is missing or the Kafka Streams `state` is not `RUNNING`.
```

> If no topics are available, the `available_topics` key will not be present in `data` of `Kafka Streams topic(s) health check`.

Reviewer: ```suggestion
If no topics are available, the `available_topics` key will not be present in the `data` field of the `Kafka Streams topics health check`.
```

> As well as if no topics are missing, the `missing_topics` key will not be present in `data` of `Kafka Streams topic(s) health check`.

Reviewer: ```suggestion
As well as if no topics are missing, the `missing_topics` key will not be present in the `data` field of the `Kafka Streams topics health check`.
```

> You can of course disable the health check of `quarkus-kafka-streams` extension by setting the `quarkus.kafka-streams.health.enabled` property to `false`.

Reviewer: ```suggestion
You can of course disable the health check of the `quarkus-kafka-streams` extension by setting the `quarkus.kafka-streams.health.enabled` property to `false` in your `application.properties`.
```

> Obviously you can create your liveness and readyness probes based on the respective endpoints `/health/live` and `/health/ready`.

Reviewer: ```suggestion
Obviously you can create your liveness and readiness probes based on the respective endpoints `/health/live` and `/health/ready`.
```

=== Liveness health check

> Here is an example of liveness:

Reviewer: ```suggestion
Here is an example of the liveness check:
```

```
curl -i http://aggregator:8080/health/live

HTTP/1.1 503 Service Unavailable
content-type: application/json; charset=UTF-8
content-length: 225

{
    "status": "DOWN",
    "checks": [
        {
            "name": "Kafka Streams state health check",
            "status": "DOWN",
            "data": {
                "state": "CREATED"
            }
        }
    ]
}
```

> The `state` is coming from `KafkaStreams.State` enum.

Reviewer: ```suggestion
The `state` is coming from the `KafkaStreams.State` enum.
```

> === Readyness health check

Reviewer: ```suggestion
=== Readiness health check
```

> Here is an example of readyness:

Reviewer: ```suggestion
Here is an example of the readiness check:
```

```
curl -i http://aggregator:8080/health/ready

HTTP/1.1 503 Service Unavailable
content-type: application/json; charset=UTF-8
content-length: 265

{
    "status": "DOWN",
    "checks": [
        {

>             "name": "Kafka Streams topic(s) health check",

Reviewer: ```suggestion
            "name": "Kafka Streams topics health check",
```

Now start Docker Compose as described above
(don't forget to rebuild the container images).

== Kafka Streams Health Check

If you are using the `quarkus-smallrye-health` extension, `quarkus-kafka-streams` will automatically add a readiness health check to validate that all topics declared in the `quarkus.kafka-streams.topics` property are created.

So when you access the `/health/ready` endpoint of your application you will have information about the available and missing topics.

> This is an example of `DOWN` status:

Reviewer 2: Very cool idea to add the absence/presence of the input topics to the health check 👍 !

In [my post](https://quarkus.io/blog/kafka-streams-applications-with-quarkus-and-microprofile/#healthchecks) from some time ago I did something slightly different: querying the stream app state:

```
public HealthCheckResponse call() {
        return HealthCheckResponse.named("My Pipeline")
                .state(streams.state().isRunning())
                .build();
    }
```

Should we have both?

Author: I don't realy know if we should or not, the decision is yours. But we technically could.
Would you like I try to add the liveness of your blog post to get a view of the result ?

Reviewer 2: I'd like to see the original check being added. It will tell you that the app is not healthy e.g. when receiving a poisonous message.

Author: Ok, I added your Liveness in addition to my Readiness. And I think the result is pretty good and meaningfull. Indeed if we only have the Liveness we don't really know why the streams is in `CREATED` state.

I also added the `streams.state()` in `data` of `HealthCheckResponse` to be more precise about why the streams is not running.

> Should we have both?

So my opinion is we really should have both ;)

http://localhost:9081/health
```
{
    "status": "DOWN",
    "checks": [
        {
            "name": "Kafka Streams state check",
            "status": "DOWN",
            "data": {
                "state": "CREATED"
            }
        },
        {
            "name": "Kafka Streams topic(s) check",
            "status": "DOWN",
            "data": {
                "Available topic(s)": "weather-stations,temperature-values",
                "Missing topic(s)": "missing-topic"
            }
        }
    ]
}
```

http://localhost:9081/health/ready
```
{
    "status": "DOWN",
    "checks": [
        {
            "name": "Kafka Streams topic(s) check",
            "status": "DOWN",
            "data": {
                "Available topic(s)": "weather-stations,temperature-values",
                "Missing topic(s)": "missing-topic"
            }
        }
    ]
}
```

http://localhost:9081/health/live
```
{
    "status": "DOWN",
    "checks": [
        {
            "name": "Kafka Streams state check",
            "status": "DOWN",
            "data": {
                "state": "CREATED"
            }
        }
    ]
}
```

```
{
    "name": "Kafka Streams health check",
    "status": "DOWN",
    "data": {
        "Available topic(s)": "weather-stations,temperature-values",
        "Missing topic(s)": "hygrometry-values"
    }
}
```
If no topics are available, the `Available topic(s)` key will not be present in `data`.

> As well as if no topics are missing, the `Missing topic(s)` key will not be present in `data` and the `status` will be `UP`.

Reviewer 3: > If no topics ..
I think this could be removed as it didn't add any usefull information.

Maybe this can be rephrased with something like:
> The check will be DOWN as soon as one topic is missing

Author: The meaning of these 2 sentences is : if the list of topics is empty, then the corresponding key is not in the body.

I precise this kind of information to ops guys that would like to set a kind of monitoring that parse the body of readiness.

Maybe the "and the `status` will be `UP`" part can be replace by you proposal before these two lines ?

Now start Docker Compose as described above
(don't forget to rebuild the container images).

== Kafka Streams Health Check

If you are using the `quarkus-smallrye-health` extension, `quarkus-kafka-streams` will automatically add:

* a readiness health check to validate that all topics declared in the `quarkus.kafka-streams.topics` property are created,
* a liveness health check based on the Kafka Streams state.

So when you access the `/health` endpoint of your application you will have information about the state of the KStreams and the available and/or missing topics.

This is an example of `DOWN` status:
[source, subs=attributes+]
----

> http http://aggregator:8080/health

Reviewer 3: Please use  ̀curl` example as it's more common than `http`

HTTP/1.1 503 Service Unavailable
content-length: 490
content-type: application/json; charset=UTF-8

{

>     "status": "DOWN",

Reviewer 2: So I'm still a bit skeptical about mixing these two thins in one health check. I will effectively make it impossible to set up a K8s liveness probe for restarting the service after a stream processing error, as also non-existing topics may trigger the restart. Would it be possible to set up two distinct health checks perhaps?

Author: hmmm... I'm affraid I don't understand what you mean.
If you base your liveness probe on `/health/live` then K8S will only restart the pod if streams state is down (a stream processing error for instance) not when topics are non-existing.

Reviewer 2: Ah, ok. Sorry, that wasn't apparent to me, if the two checks can be queried independently, then all is good.

Author: Maybe the adoc is not clear enough ?

Reviewer 2: Yes, it might be a good idea to show how the two can be queried independently.
